AudioPanel: Refactor to use Config system.

TryTwo opened the pull request:
Updated the image to the current version.

Placing everything under the config system will make things easier to use and re-arrange long term, especially if someone wants to change up the UI.  Also, it will allow Local game-ini options to be bolded and possibly set directly.

``AudioCommon::UpdateSoundStream(Core::System::GetInstance())`` I assumed this didn't need to be called for things that can only be changed before running a game.  If an audio update is actually required for such things, I would trigger it on OnEmulationStateChanged.

![image](https://github.com/user-attachments/assets/8bface4d-7ee0-489f-ba5b-8372c330a2f1)

mbc07 reviewed (changes requested):
UI wise, the immediately apparent visual changes seems to be the DPL2 quality setting (which is changed from a slider to a dropdown menu) and the WASAPI backend latency setting (which is changed from an text input to a slider). Those changes are fine to me, however, in its current state, there are other issues with this PR:

* The new DPL2 quality setting dropdown seems broken. It's always in disabled state, even after selecting one of the DSP LLE engines, the Cubeb audio backend and checking "Dolby Pro Logic 2 Decoder"
* There's a small visual bug with the latency value of the Audio Stretching buffer slider and the Volume slider, they only show up after interacting with the slider instead of being always visible (like in master)
* As a side effect of the refactor, this PR introduces the custom tooltips to the controls in this page, but they're very inconsistent:
  * All custom tooltips introduced lacks the description text (as well as the usual "if unsure, select X" emphasis at the end)
  * DSP HLE and DSP LLE Recompiler radio boxes have custom tooltips, but DSP LLE Interpreter does not. For consistency, please add a custom tooltip to it too.
  * Backend, Device (shown only when WASAPI is selected) and Volume lacks custom tooltips too.
  * Backend latency, audio stretching buffer and DPL2 decoder quality has legacy tooltips only, please convert them to the custom tooltips.
  * Enable Audio Stretching and DPL2 Decoder has both legacy and custom tooltips, please drop the legacy ones and move their text to the custom tooltips

I know I'm asking a lot here and some of that likely are outside of the scope of this PR (especially adding description text for the tooltips), but I can't unsee the several inconsistencies this PR currently introduces. 

With that said, if you don't want to address all of that in this PR, I'd like to see at least the controls that currently have both legacy and custom tooltips converted to use custom tooltips only, DSP LLE interpreter radio box getting a custom tooltip for consistency, the visual bug with the audio stretching/volume sliders and the broken decoder quality dropdown fixed before giving an approval...

TryTwo commented:
Thanks, that's exactly what I needed.  I think I addressed everything. I BS'd some of the descriptions, because I wasn't sure about them, but  I wanted to get it all setup for someone else to write better ones.  I can comment out the descriptions before it's merged if needed.

I wasn't sure what you meant by some things having both legacy and custom tooltips, but I think it's sorted now.

mbc07 commented:
> I wasn't sure what you meant by some things having both legacy and custom tooltips, but I think it's sorted now.

By legacy tooltips I was referring to the system ones, set via Qt's `element->setTooltip()` method. By custom ones I was referring to Dolphin's custom implementation. You indeed took care of all of them, thanks. I'm finishing a list of text changes for the descriptions you've added (so we have an MVP) and then this is good to go from me...

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+      tr(TR_AUDIO_BACKEND_DESCRIPTION)
+          .arg(QString::fromStdString(AudioCommon::GetDefaultSoundBackend())));
+
+  m_dolby_pro_logic->SetTitle(tr("Dolby Pro Logic II"));
```
`Dolby Pro Logic II Decoder`

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+  m_dolby_pro_logic->SetTitle(tr("Dolby Pro Logic II"));
+  m_dolby_pro_logic->SetDescription(tr(TR_DOLBY_DESCRIPTION));
+
+  m_dolby_quality_combo->SetTitle(tr("Dolby Quality"));
```
`Decoding Quality`

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
-{
-  switch (value)
+#ifdef _WIN32
+  m_wasapi_device_combo->SetTitle(tr("WASAPI Device"));
```
`Output Device`

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -80,79 +82,59 @@ void AudioPane::CreateWidgets()
+                                                         QString::fromStdString(string)});
+  }
+
   m_wasapi_device_label = new QLabel(tr("Device:"));
```
`Output Device:`

(since you're already refactoring the panel, I'll ask for this too)

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+  m_stretching_enable->SetTitle(tr("Audio Stretching"));
+  m_stretching_enable->SetDescription(tr(TR_STRETCH_SLIDER_DESCRIPTION));
+
+  m_stretching_buffer_slider->SetTitle(tr("Audio Stretching Slider"));
```
`Buffer Size`

mbc07 reviewed (changes requested):
And that's it. I know it's a lot, but those changes ensures the newly introduced tooltips keep consistent with the wording and style conventions of the other tooltips that already exist in Dolphin (used the Graphics window as a base, which arguably hosts the most polished tooltips Dolphin currently has).

I don't think I have anything else to ask, once those are addressed, this PR LGTM.

TryTwo commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -80,79 +82,59 @@ void AudioPane::CreateWidgets()
-  dolby_quality_layout->addWidget(m_dolby_quality_highest_label);
+  QStringList quality_options{
+      tr("Lowest Quality (528), Fastest (~10 ms)"), tr("Low Qualiity (1024), Fast (~20 ms)"),
+      tr("High Quality (2048), Slow (~40 ms)"), tr("Highest Quality (4096). Slowest (~80 ms)")};
```
I'm ok with changing it. Because we're listing two things, I think it can get a little confusing.  I was trying to use Fast and Slow to keep it clear.  I don't mind dropping any/all the numbers, but I'm not sure what is expected.

mbc07 replied:
My pet peeve with this is that a lower/bigger latency isn't necessarily correlated to a increase/decrease in speed, as implied by the use of Fastest/Fast/Slow/Slowest labels. I'll leave this to you, but it bugs me a little...

TryTwo commented:
Thanks for helping me with this.

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+#endif
+
+  m_stretching_enable->SetTitle(tr("Audio Stretching"));
+  m_stretching_enable->SetDescription(tr(TR_STRETCH_SLIDER_DESCRIPTION));
```
This one got mixed. Should be `m_stretching_enable->SetDescription(tr(TR_STRETCH_ENABLE_DESCRIPTION));`

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+  static const char TR_INTERPRETER_DESCRIPTION[] =
+      QT_TR_NOOP("Low Level Emulation of the DSP, via a interpreter. Slowest, for debugging "
+                 "purposes only.<br><br><dolphin_emphasis>Not recommended, consider using DSP HLE "
+                 "or DSP LLE Recompiler instead.</dolphin_emphasis>");
```
Small typo I introduced in my previous review, edited the comment but it apparently didn't went through (via a interpreter => via an interpreter): 

`Low Level Emulation of the DSP, via an interpreter. Slowest, for debugging purposes only.<br><br><dolphin_emphasis>Not recommended, consider using DSP HLE or DSP LLE Recompiler instead.</dolphin_emphasis>`

mbc07 reviewed (changes requested):
Almost there...

JosJuice commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+  static const char TR_LLE_DESCRIPTION[] = QT_TR_NOOP(
+      "DSP Low Level Emulation. Slower, but more accurate method to play audio. Enables Dolby "
+      "features on certain backends. <br><br><dolphin_emphasis> If unsure, select &quot;DSP HLE"
+      "&quot;.</dolphin_emphasis>");
```
No comma splice between "accurate" and "also", please.

JosJuice commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -386,68 +232,100 @@ void AudioPane::OnEmulationStateChanged(bool running)
+      "&quot;.</dolphin_emphasis>");
+  static const char TR_INTERPRETER_DESCRIPTION[] = QT_TR_NOOP(
+      "DSP INterpreter. The slowest method, used for debugging. Not recommended."
+      "<br><br><dolphin_emphasis> If unsure, select &quot;DSP HLE&quot;.</dolphin_emphasis>");
```
There's a comma splice here too, after "not recommended".

TryTwo commented:
Hmm you're right about the latency vs fast.

mbc07 reviewed (changes requested):
There's a single title change that I missed before, then this LGTM (I swear this time it's the last request from me). Also

> For me the DSP LLE tooltip is squished horizontally. Not sure what causes that. It seems random.

If I recall correctly, it's related to the lack/presence of certain tags in the tooltip, which has a side effect of influencing the text wrapping, but I'm not totally sure. Not a blocker, IMHO

TryTwo commented:
It seems the tooltip is squishing because it has to wrap the tip to a second line.  If I use a <br> before it gets too long, it won't break. Using one after it's too long still squishes.

Since the wrapping is probably resolution dependent, I  guess it's not a good idea to put a <br> in before the last sentence.

TryTwo commented:
Updated for Granule Audio commits.

Changed DSP selection to a combo box, which works better with how the settings are stored.  Probably need to work on the text and descriptions for it. @mbc07 do you want to help with that?

mbc07 commented:
I'm a bit busy at the moment, but I'll check this PR soon...

TryTwo commented:
![0hQTRRIxCX](https://github.com/user-attachments/assets/d25ee2dd-1c47-427b-8b92-49e9d1a727cd)

mbc07 reviewed (changes requested):
I was hoping we could keep the radio boxes, but if they have to go, I think the dropdown that's going to replace them should be consistent with other similar dropdowns already present in Dolphin. Here's a quick mockup:

![image](https://github.com/user-attachments/assets/f2c9c504-4197-4c5f-9cd8-1866c0612907)

Requested changes (in comparison to the current version of the PR):
* Rename the group box containing the dropdown from "DSP Emulator Settings" to "DSP Options";
* Add a text label with the setting name (DSP Emulator Settings) to the left of the dropdown.

And I think that's it. Feels a bit odd having a group box with a single setting, but design-wise it's very consistent with the "CPU Options" group box from the Advanced tab, which has a similar purpose.

Regarding the tooltips, their current iteration LGTM, but I noticed that the tooltip from the "Mute When Disabling Speed Limit" checkbox is lacking a "If unsure, leave this unchecked" in `<dolphin_emphasis>` at the end. I don't remember if that was already the case when I first reviewed this PR (maybe not, I don't recall this checkbox already existing at that point), but please fix that too...

(also sorry for the delay, lately I've been busier than the usual)

mbc07 commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -361,76 +203,101 @@ void AudioPane::OnBackendChanged()
+      "Repeat existing audio during lag spikes to prevent stuttering.<br><br><dolphin_emphasis>If "
+      "unsure, leave this checked.</dolphin_emphasis>");
+  static const char TR_SPEED_UP_MUTE_DESCRIPTION[] =
+      QT_TR_NOOP("Mutes the audio when overriding the emulation speed limit(default hotkey: Tab). "
```
A space is missing between "speed limit" and the parenthesis

mbc07 reviewed (approved):
The current iteration of this PR LGTM after fixing the small typo...

TryTwo commented:
@sepalani I think this would be a quick code review if you want to take a look?  It's been tested well on windows at least, but I don't think there's much of a platform difference.

sepalani commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -46,29 +48,28 @@ AudioPane::AudioPane()
-  m_dsp_lle = new QRadioButton(tr("DSP LLE Recompiler (slow)"));
-  m_dsp_interpreter = new QRadioButton(tr("DSP LLE Interpreter (very slow)"));
+  QLabel* dsp_combo_label = new QLabel(tr("DSP Emulation Engine:"));
+  m_dsp_combo = new ConfigComplexChoice(Config::MAIN_DSP_HLE, Config::MAIN_DSP_JIT);
+  m_dsp_combo->Add(tr("HLE (recommended)"), true, true);
+  m_dsp_combo->Add(tr("LLE Recompiler (slow)"), false, true);
+  m_dsp_combo->Add(tr("LLE Interpreter (very slow)"), false, false);
+  m_dsp_combo->Refresh();
```
When I first tested this PR, none of them were selected:
![image](https://github.com/user-attachments/assets/854555ea-1b84-4b78-b656-58e68e847a80)

I don't recall having changed my audio config beside switching between HLE/LLE depending on the game I start. I suppose there might/will be users with the same issue. One way to reproduce it, is to set `MAIN_DSP_HLE = True` and `Config::MAIN_DSP_JIT = False`, for instance:
```INI
[DSP]
WASAPIDevice = 
EnableJIT = False
[Core]
DSPHLE = True
```

TryTwo replied:
I added a new commit to set a default index if there is no valid one, which fixes this (as there is only one invalid state).

sepalani commented on Source/Core/DolphinQt/Settings/AudioPane.cpp:
```diff
@@ -24,19 +20,25 @@
 #include "DolphinQt/Config/SettingsWindow.h"
 #include "DolphinQt/Settings.h"
 
+constexpr int DSP_HLE_INDEX = 0;
```
Can't you get the same result using `Config::Get(MAIN_DSP_HLE)` instead? When I first tested this PR, the `enabled` checks weren't working properly (e.g. Dolby Pro Logic II enabled when the combobox was blank).

TryTwo commented:
Nice finds. 
```
[DSP]
EnableJIT = False
[Core]
DSPHLE = True
```
I believe the above is technically an unexpected global state, because normally if you do DSPHLE =  false in a game ini, then you expect it to give you JIT. If the global is the above state, it would give you interpreter instead. The old option format allowed this to happen though. @JMC47 Is this correct?

I added a new commit to set a default index if there is no valid one, which fixes this (as there is only one invalid state).

sepalani reviewed (approved):
Tested on Windows, LGTM, I can't reproduce my previous issue anymore.

Unfortunately, I'm unable to test the Dolby quality feature.

xujibbs commented:
when update to this version and after, "No Audio Output" and "WASAPI (Exclusive Mode)" in MainSettings.h is untranslatable.
